File: bindings/rust/src/runtime/decode.rs
//! The one analytic row decoder, driven by the generated schema table.
//!
//! Seventeen verbs return the same self-describing row, so there is exactly one
//! decoder here rather than seventeen result structs drifting independently.
//! It walks `SCHEMAS[schema_id - 1]` *positionally* over the value array, which
//! is what makes a verb that reuses an existing schema cost zero new code.
//!
//! Three properties the row model exists to preserve:
//!
//! * **Absent is not zero.** `distance = 0.0` means *identical*, so a cleared
//!   presence bit becomes `None`, never a sentinel.
//! * **Unknown stays unknown.** `[row_enums]` is append-only; an ordinal past
//!   this build's table is carried verbatim in a [`Variant`] whose `name()` is
//!   `None`, rather than clamped to the nearest variant we happen to know.
//! * **A row that exists is valid.** Construction runs the recursive check in
//!   [`super::verify`], so every accessor below is infallible.
//!
//! # Lifetimes
//!
//! A [`Row`] is a borrowed view: it holds the row header by value (four words)
//! and a `PhantomData` tying it to the arena its texts and nested rows live in.
//! For a native answer that arena is the `irgx_rows` cursor, which the C ABI
//! guarantees stays valid until close — so batches may be held alongside each
//! other without copying. [`Row::to_owned`] is the explicit exit from the
//! borrow into an [`OwnedRow`] that outlives the cursor.

use std::marker::PhantomData;

use super::cell::{OwnedRow, OwnedValue, RowSeq, Texts, Value, borrow, read};
use super::{Result, sys};
use crate::contract::Variant;
use crate::contract::schema::{FieldDef, SCHEMAS, SchemaDef};

/// The declared schema for a wire id, or `None` for one this build predates.
#[must_use]
pub fn schema_for(id: u32) -> Option<&'static SchemaDef> {
    SCHEMAS.get(usize::try_from(id).ok()?.checked_sub(1)?)
}

/// One analytic result row, borrowed from the arena it was materialized into.
///
/// The row is four words plus a lifetime marker, so it is [`Copy`] and cheap to
/// pass around; everything it points at lives in the arena, which is why the
/// lifetime is not decorative.
#[derive(Debug, Clone, Copy)]
pub struct Row<'a> {
    schema: &'static SchemaDef,
    cells: Cells<'a>,
}

#[derive(Debug, Clone, Copy)]
enum Cells<'a> {
    /// A native answer: the raw value array plus the presence mask.
    Wire {
        present: u64,
        values: *const sys::Value,
        nvalues: usize,
        arena: PhantomData<&'a ()>,
    },
    /// A subprocess answer, already lowered into owned cells.
    Owned(&'a [Option<OwnedValue>]),
}

impl<'a> Row<'a> {
    /// Decode and validate one wire row.
    ///
    /// # Errors
    /// [`super::Error::Decode`] for an unknown `schema_id`, a value whose tag
    /// disagrees with its declaration, or text that is not UTF-8 — each named
    /// with the schema and field so a drift is diagnosable, never silent.
    ///
    /// # Safety
    /// `raw` and everything it points at must be valid for `'a`.
    pub(crate) unsafe fn from_wire(raw: &sys::Row) -> Result<Self> {
        let row = Self::wire_unchecked(raw);
        row.validate()?;
        Ok(row)
    }

    /// The same view without the check — for a row whose parent already
    /// validated it recursively.
    pub(super) fn wire_unchecked(raw: &sys::Row) -> Self {
        Self {
            // A row past this build's table validates as unknown before any
            // accessor runs; the placeholder keeps the type total meanwhile.
            schema: schema_for(raw.schema_id).unwrap_or(UNKNOWN_SCHEMA),
            cells: Cells::Wire {
                present: raw.present,
                values: raw.values,
                nvalues: raw.nvalues as usize,
                arena: PhantomData,
            },
        }
    }

    /// Borrow already-owned cells through the wire accessors. `None` when the
    /// id names a schema this build does not declare.
    pub(super) fn over(schema_id: u32, cells: &'a [Option<OwnedValue>]) -> Option<Self> {
        Some(Self {
            schema: schema_for(schema_id)?,
            cells: Cells::Owned(cells),
        })
    }

    /// The wire mask, pointer, and arity — `None` for an owned row, whose cells
    /// were validated positionally when they were built.
    pub(super) fn wire_cells(self) -> Option<(u64, *const sys::Value, usize)> {
        match self.cells {
            Cells::Wire {
                present,
                values,
                nvalues,
                ..
            } => Some((present, values, nvalues)),
            Cells::Owned(_) => None,
        }
    }

    /// The `[row_schemas]` id this row was decoded against.
    #[must_use]
    pub fn schema_id(self) -> u32 {
        self.schema.id
    }

    /// The `[row_schemas]` key (`"similar"`, `"blast"`, …).
    #[must_use]
    pub fn schema_name(self) -> &'static str {
        self.schema.name
    }

    /// The declared field list, in wire order.
    #[must_use]
    pub fn fields(self) -> &'static [FieldDef] {
        self.schema.fields
    }

    /// The value of declared field `i`, or `None` when the presence mask says
    /// it is absent (which is *not* the same as zero).
    #[must_use]
    pub fn get(self, i: usize) -> Option<Value<'a>> {
        let field = self.schema.fields.get(i)?;
        match self.cells {
            Cells::Wire {
                present,
                values,
                nvalues,
                ..
            } => {
                if i >= nvalues || present & (1u64 << i) == 0 {
                    return None;
                }
                Some(read(field, unsafe { &*values.add(i) }))
            },
            Cells::Owned(cells) => cells.get(i)?.as_ref().map(borrow),
        }
    }

    /// The value of the declared field named `name`.
    #[must_use]
    pub fn field(self, name: &str) -> Option<Value<'a>> {
        let i = self.schema.fields.iter().position(|f| f.name == name)?;
        self.get(i)
    }

    /// Field `name` as text.
    #[must_use]
    pub fn text(self, name: &str) -> Option<&'a str> {
        self.field(name)?.as_str()
    }

    /// Field `name` as an integer.
    #[must_use]
    pub fn int(self, name: &str) -> Option<i64> {
        self.field(name)?.as_i64()
    }

    /// Field `name` as a real.
    #[must_use]
    pub fn real(self, name: &str) -> Option<f64> {
        self.field(name)?.as_f64()
    }

    /// Field `name` as a boolean.
    #[must_use]
    pub fn flag(self, name: &str) -> Option<bool> {
        self.field(name)?.as_bool()
    }

    /// Field `name` as an enum ordinal (unresolved — see [`Variant::name`]).
    #[must_use]
    pub fn variant(self, name: &str) -> Option<Variant> {
        self.field(name)?.as_variant()
    }

    /// Field `name` as a string list.
    #[must_use]
    pub fn texts(self, name: &str) -> Option<Texts<'a>> {
        self.field(name)?.as_texts()
    }

    /// Field `name` as nested rows.
    #[must_use]
    pub fn rows(self, name: &str) -> Option<RowSeq<'a>> {
        self.field(name)?.as_rows()
    }

    /// Iterate `(declared name, value)` for every present field.
    pub fn iter(self) -> impl Iterator<Item = (&'static str, Value<'a>)> {
        self.schema
            .fields
            .iter()
            .enumerate()
            .filter_map(move |(i, f)| self.get(i).map(|v| (f.name, v)))
    }

    /// Copy the row out of the arena so it can outlive the cursor.
    #[must_use]
    pub fn to_owned(self) -> OwnedRow {
        OwnedRow {
            schema_id: self.schema.id,
            values: (0..self.schema.fields.len())
                .map(|i| self.get(i).map(Value::to_owned))
                .collect(),
        }
    }
}

/// Stand-in for a row whose schema this build predates; validation rejects it
/// before any accessor can read a field it cannot name.
static UNKNOWN_SCHEMA: &SchemaDef = &SchemaDef {
    id: 0,
    name: "unknown",
    fields: &[],
};

#[cfg(test)]
mod tests {
    //! Expectations come from the contract — `SCHEMAS`, `ENUMS`, and the field
    //! positions read out of them — never from a run of this decoder.

    use super::super::fixture::{
        self, all, enumeration, field_at, int, real, row, schema_id, text,
    };
    use super::*;
    use crate::contract::{Channel, Grade};

    #[test]
    fn absent_is_not_zero() {
        // `similar.distance` is 0.0 for a byte-identical file, so the decoder
        // must distinguish "no distance reported" from "distance is zero" — the
        // single reason the presence mask exists.
        let id = schema_id("similar");
        let d = field_at(id, "distance");
        let values = [text("a.rs"), real(0.0), enumeration(0), enumeration(0)];

        let full = unsafe { Row::from_wire(&row(id, all(values.len()), &values)) };
        assert_eq!(full.expect("valid row").real("distance"), Some(0.0));

        let masked = unsafe { Row::from_wire(&row(id, all(values.len()) & !(1 << d), &values)) }
            .expect("valid row");
        assert_eq!(masked.real("distance"), None);
        // Absence is per field: the neighbors still read.
        assert_eq!(masked.text("path"), Some("a.rs"));
    }

    #[test]
    fn unknown_enum_ordinal_stays_unknown() {
        // `[row_enums]` is append-only, so a newer engine can report an ordinal
        // past this build's table. It must survive verbatim rather than being
        // clamped to the nearest variant we happen to know.
        let id = schema_id("similar");
        let beyond = i64::try_from(Grade::ALL.len()).unwrap_or(i64::MAX) + 3;
        let values = [text("a.rs"), real(0.5), enumeration(beyond), enumeration(0)];
        let decoded =
            unsafe { Row::from_wire(&row(id, all(values.len()), &values)) }.expect("valid row");

        let grade = decoded.variant("grade").expect("present");
        assert_eq!(grade.ordinal, beyond);
        assert_eq!(grade.name(), None, "an unnameable ordinal is not guessable");
        assert_eq!(Grade::from_variant(grade), None);
        assert_eq!(grade.to_string(), format!("grade?{beyond}"));
        // The neighboring enum, in range, still resolves.
        assert_eq!(
            decoded.variant("channel").and_then(Channel::from_variant),
            Some(Channel::Copies)
        );
    }

    #[test]
    fn an_enum_resolves_against_its_own_vocabulary_not_its_position() {
        // `similar` carries two enums back to back; reading either against the
        // other's table would silently mislabel every kinship row.
        let id = schema_id("similar");
        let values = [text("a.rs"), real(0.0), enumeration(3), enumeration(1)];
        let decoded = unsafe { Row::from_wire(&row(id, all(4), &values)) }.expect("valid row");
        assert_eq!(
            decoded.variant("grade").and_then(Grade::from_variant),
            Some(Grade::Strong)
        );
        assert_eq!(
            decoded.variant("channel").and_then(Channel::from_variant),
            Some(Channel::Twins)
        );
        // A grade ordinal read as a channel is refused, not coerced.
        let grade = decoded.variant("grade").expect("present");
        assert_eq!(Channel::from_variant(grade), None);
    }

    #[test]
    fn nested_rows_recurse_and_carry_their_own_mask() {
        let region = schema_id("region");
        let family = schema_id("family");
        let head = field_at(region, "headline");

        let member_values = [text("a.rs"), int(10), int(20), text("fn a")];
        let members = [
            row(region, all(member_values.len()), &member_values),
            // The second member withholds the optional headline.
            row(
                region,
                all(member_values.len()) & !(1 << head),
                &member_values,
            ),
        ];
        let values = [
            int(1),
            enumeration(0),
            enumeration(0),
            real(0.2),
            int(80),
            real(0.9),
            fixture::rows(&members),
        ];
        let decoded =
            unsafe { Row::from_wire(&row(family, all(values.len()), &values)) }.expect("valid row");

        let nested = decoded.rows("members").expect("members present");
        assert_eq!(nested.len(), 2);
        let first = nested.get(0).expect("first member");
        assert_eq!(first.schema_name(), "region");
        assert_eq!(first.text("path"), Some("a.rs"));
        assert_eq!(first.int("line_start"), Some(10));
        assert_eq!(first.text("headline"), Some("fn a"));
        assert_eq!(
            nested.get(1).and_then(|r| r.text("headline")),
            None,
            "the nested presence mask is honored per nested row"
        );
    }

    #[test]
    fn a_string_list_reads_as_many_borrowed_strs() {
        let id = schema_id("cluster");
        let paths = [
            sys::Text {
                ptr: "a.rs".as_ptr(),
                len: 4,
            },
            sys::Text {
                ptr: "b.rs".as_ptr(),
                len: 4,
            },
        ];
        let values = [fixture::texts(&paths), real(0.1), enumeration(3)];
        let decoded = unsafe { Row::from_wire(&row(id, all(3), &values)) }.expect("valid row");
        let list = decoded.texts("paths").expect("paths present");
        assert_eq!(list.len(), 2);
        assert_eq!(list.iter().collect::<Vec<_>>(), ["a.rs", "b.rs"]);
    }

    #[test]
    fn a_short_value_array_reads_as_absent_rather_than_out_of_bounds() {
        // An older engine may stop short of a field a newer contract appends.
        // Only the overlap is ours to read, and the tail is absent, not zero.
        let id = schema_id("similar");
        let values = [text("a.rs"), real(0.25)];
        let decoded =
            unsafe { Row::from_wire(&row(id, all(4), &values)) }.expect("a prefix is decodable");
        assert_eq!(decoded.real("distance"), Some(0.25));
        assert_eq!(decoded.variant("grade"), None);
    }

    #[test]
    fn iter_yields_exactly_the_present_fields_in_declared_order() {
        let id = schema_id("similar");
        let values = [text("a.rs"), real(0.25), enumeration(3), enumeration(0)];
        let decoded = unsafe { Row::from_wire(&row(id, 0b0101, &values)) }.expect("valid row");
        let names: Vec<&str> = decoded.iter().map(|(n, _)| n).collect();
        assert_eq!(names, ["path", "grade"]);
    }

    #[test]
    fn owned_rows_read_through_the_same_accessors() {
        let id = schema_id("similar");
        let values = [text("a.rs"), real(0.0), enumeration(3), enumeration(1)];
        let borrowed = unsafe { Row::from_wire(&row(id, 0b1111, &values)) }.expect("valid row");
        let owned = borrowed.to_owned();
        let view = owned.view().expect("known schema");
        assert_eq!(view.text("path"), borrowed.text("path"));
        assert_eq!(view.real("distance"), Some(0.0));
        assert_eq!(
            view.variant("grade").and_then(Grade::from_variant),
            Some(Grade::Strong)
        );
    }

    #[test]
    fn an_owned_row_keeps_absence_through_the_round_trip() {
        let id = schema_id("similar");
        let d = field_at(id, "distance");
        let values = [text("a.rs"), real(0.0), enumeration(0), enumeration(0)];
        let borrowed = unsafe { Row::from_wire(&row(id, all(values.len()) & !(1 << d), &values)) }
            .expect("valid row");
        let owned = borrowed.to_owned();
        assert_eq!(owned.values[d], None);
        assert_eq!(owned.view().and_then(|v| v.real("distance")), None);
    }

    #[test]
    fn owning_a_nested_row_copies_the_whole_tree() {
        let region = schema_id("region");
        let family = schema_id("family");
        let member_values = [text("a.rs"), int(10), int(20), text("fn a")];
        let members = [row(region, all(member_values.len()), &member_values)];
        let values = [
            int(1),
            enumeration(0),
            enumeration(0),
            real(0.2),
            int(80),
            real(0.9),
            fixture::rows(&members),
        ];
        let owned = unsafe { Row::from_wire(&row(family, all(values.len()), &values)) }
            .expect("valid row")
            .to_owned();
        let nested = owned
            .view()
            .and_then(|r| r.rows("members"))
            .expect("members survive the copy");
        assert_eq!(nested.get(0).and_then(|r| r.text("headline")), Some("fn a"));
    }
}

File: bindings/go/runtime/decode.go
// The cold tier's raising half: a child's NDJSON back into rows of the verb's
// declared schema. It is the mirror of plan.go — that file spells a query as argv,
// this one reads the answer — and the two are kept apart because a CLI's row
// vocabulary drifts on a different clock than its flag vocabulary.

package runtime

import (
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"slices"
	"strconv"
	"strings"

	"github.com/The-Billy-Company/irregex/bindings/go/v2/analytic"
)

// jsonKeys maps a schema field onto the JSON path(s) a CLI row spells it with,
// first present winning. Most fields need no entry — the names already agree —
// so this table is exactly the divergences, and a row field with no entry and no
// matching key decodes as ABSENT rather than as a zero.
var jsonKeys = map[string][]string{
	"similar.path":                {"unit", "path"},
	"recalled.path":               {"unit", "path"},
	"echo.byte_distance":          {"bytes"},
	"echo.structure_distance":     {"structure"},
	"concept.byte_distance":       {"bytes"},
	"concept.structure_distance":  {"structure"},
	"cluster.paths":               {"members", "paths"},
	"cluster.max_distance":        {"distance", "max_distance"},
	"family.edge":                 {"echo", "distance"},
	"family.score":                {"distance", "echo"},
	"distinct.member":             {"unit", "member"},
	"distinct.byte_distance":      {"bytes"},
	"distinct.structure_distance": {"structure"},
	"blast.symbol":                {"seed.symbol"},
	"blast.kind":                  {"seed.kind"},
	"blast.definitions":           {"seed.def"},
	"blast.dependents":            {"direct.dependents"},
	"blast.dependencies":          {"direct.dependencies"},
	"blast.twins":                 {"tangential.twins"},
	"blast.ripple":                {"tangential.ripple"},
	"blast.omitted":               {"stats.omitted"},
	"reference.enclosing":         {"in", "enclosing"},
	"reference.defines":           {"use", "defines"},
	"region.line_start":           {"line", "line_start"},
	"region.line_end":             {"line_end", "line"},
}

// rowsPerObject decodes one JSON object per line into one row of schema.
func rowsPerObject(schema uint32, stdout string) ([]Row, error) {
	objects, err := ndjson(stdout)
	if err != nil {
		return nil, err
	}
	answer := summary(stdout, "")
	rows := make([]Row, 0, len(objects))
	for _, obj := range objects {
		row, err := rowFromJSON(schema, obj, answer)
		if err != nil {
			return nil, err
		}
		rows = append(rows, row)
	}
	return rows, nil
}

// headThenRows decodes the one-row verbs whose CLI prints a summary object and
// then the members of its nested field (quote's phrases).
func headThenRows(schema uint32, stdout string) ([]Row, error) {
	objects, err := ndjson(stdout)
	if err != nil || len(objects) == 0 {
		return nil, err
	}
	def, ok := analytic.Schema(schema)
	if !ok {
		return nil, fmt.Errorf("irregex: unknown row schema id %d", schema)
	}
	head, err := rowFromJSON(schema, objects[0], nil)
	if err != nil {
		return nil, err
	}
	for i, f := range def.Fields {
		if analytic.Tag(f.Tag) != analytic.TagRows {
			continue
		}
		children := make([]Row, 0, len(objects)-1)
		for _, obj := range objects[1:] {
			child, err := rowFromJSON(f.Nested, obj, nil)
			if err != nil {
				return nil, err
			}
			children = append(children, child)
		}
		head.values[i] = Nested(children)
		head.present |= 1 << uint(i)
		break
	}
	return []Row{head}, nil
}

// rankRow is one `--rank` line: index, path:line, the engine's own def/use/gen
// class, the per-file count behind a multiplication sign, then the snippet. Rank
// is the one verb whose CLI answer is a rendered view rather than NDJSON.
var rankRow = regexp.MustCompile(`^\s*\d+\.\s+(.+?):(\d+)\s+\[(def|use|gen)\]\s+\x{00d7}(\d+)\s*(.*)$`)

var rankClass = map[string]string{"def": "definition", "use": "use", "gen": "generated"}

func rankedRows(schema uint32, stdout string) ([]Row, error) {
	var rows []Row
	for line := range strings.SplitSeq(stdout, "\n") {
		m := rankRow.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		row, err := rowFromJSON(schema, map[string]any{
			"path":        m[1],
			"line_number": m[2],
			"kind":        rankClass[m[3]],
			"count":       m[4],
			"snippet":     m[5],
		}, nil)
		if err != nil {
			return nil, err
		}
		rows = append(rows, row)
	}
	return rows, nil
}

// inherited names the row fields the CLI states once per answer instead of once
// per row, mapped to the summary key that carries them. These are not guesses: the
// summary asserts the unit and channel the whole answer was computed in, and the
// row's own same-named key means something else entirely (a member path).
var inherited = map[string]string{
	"family.unit":    "unit",
	"family.channel": "channel",
	"distinct.unit":  "unit",
}

// rowFromJSON assembles one row of schema from a decoded JSON object, with answer
// the verb's summary record (nil when there is none). It walks the schema's
// declared fields — never the object's keys — so a field the tool did not report
// stays absent, and a key the contract does not declare cannot smuggle itself into
// a row.
func rowFromJSON(schema uint32, obj, answer map[string]any) (Row, error) {
	def, ok := analytic.Schema(schema)
	if !ok {
		return Row{}, fmt.Errorf("irregex: unknown row schema id %d", schema)
	}
	values, present := make([]Value, len(def.Fields)), uint64(0)
	for i, f := range def.Fields {
		raw, found := lookup(obj, answer, def.Name, f.Name)
		if !found || raw == nil {
			continue
		}
		v, ok, err := valueFromJSON(f, raw)
		if err != nil {
			return Row{}, fmt.Errorf("irregex: %s.%s: %w", def.Name, f.Name, err)
		}
		if !ok {
			continue
		}
		values[i] = v
		present |= 1 << uint(i)
	}
	return Assemble(schema, values, present)
}

func lookup(obj, answer map[string]any, schema, field string) (any, bool) {
	if key, ok := inherited[schema+"."+field]; ok {
		if v, ok := answer[key]; ok && v != nil {
			return v, true
		}
	}
	paths := jsonKeys[schema+"."+field]
	if paths == nil {
		paths = []string{field}
	}
	for _, path := range paths {
		if v, ok := walk(obj, path); ok && v != nil {
			return v, true
		}
	}
	return nil, false
}

func walk(obj map[string]any, path string) (any, bool) {
	cur := any(obj)
	for step := range strings.SplitSeq(path, ".") {
		node, ok := cur.(map[string]any)
		if !ok {
			return nil, false
		}
		if cur, ok = node[step]; !ok {
			return nil, false
		}
	}
	return cur, true
}

// valueFromJSON coerces one JSON value into its declared tag. ok is false when
// the tool reported something that is not a measurement — a non-finite number, an
// enum label this binding does not know — which is absence, not a zero.
func valueFromJSON(f analytic.FieldDef, raw any) (Value, bool, error) {
	switch analytic.Tag(f.Tag) {
	case analytic.TagText:
		return Text(text(raw)), true, nil
	case analytic.TagInt:
		n, ok := number(raw)
		return Int(int64(n)), ok, nil
	case analytic.TagFloat:
		n, ok := number(raw)
		return Float(n), ok, nil
	case analytic.TagBool:
		return Bool(truthy(raw)), true, nil
	case analytic.TagEnum:
		return enumValue(f.Nested, raw)
	case analytic.TagTexts:
		items, ok := raw.([]any)
		if !ok {
			return Value{}, false, fmt.Errorf("expected a list, got %T", raw)
		}
		out := make([]string, len(items))
		for i, item := range items {
			out[i] = text(item)
		}
		return Texts(out), true, nil
	case analytic.TagRows:
		// A single-member nested field (distinct's member and its nearest miss)
		// arrives unwrapped, so one row and a list of one are the same answer.
		items, ok := raw.([]any)
		if !ok {
			items = []any{raw}
		}
		rows := make([]Row, 0, len(items))
		for _, item := range items {
			child, err := childRow(f.Nested, item)
			if err != nil {
				return Value{}, false, err
			}
			rows = append(rows, child)
		}
		return Nested(rows), true, nil
	}
	return Value{}, false, fmt.Errorf("field declares tag %d, which this decoder does not know", f.Tag)
}

// childRow decodes one nested member. A member may arrive as an object, or — for
// the family and cluster shapes — as the string `path#Lnnn`, which lifts into the
// nested schema's own path/line fields rather than being dropped.
func childRow(schema uint32, item any) (Row, error) {
	if obj, ok := item.(map[string]any); ok {
		return rowFromJSON(schema, obj, nil)
	}
	path, line, found := strings.Cut(text(item), "#L")
	lifted := map[string]any{"path": path}
	if found {
		if n, err := strconv.ParseInt(line, 10, 64); err == nil {
			lifted["line"] = float64(n)
		}
	}
	return rowFromJSON(schema, lifted, nil)
}

func enumValue(id uint32, raw any) (Value, bool, error) {
	if label, ok := raw.(string); ok {
		ordinal, known := analytic.EnumOrdinal(id, label)
		if !known {
			// A label this table cannot name must not become a neighboring
			// ordinal; -1 is outside every variant list, so it reads back unknown.
			return EnumOf(id, -1), true, nil
		}
		return EnumOf(id, ordinal), true, nil
	}
	n, ok := number(raw)
	return EnumOf(id, int64(n)), ok, nil
}

func text(raw any) string {
	switch v := raw.(type) {
	case nil:
		return ""
	case string:
		return v
	case float64:
		return strconv.FormatFloat(v, 'g', -1, 64)
	case bool:
		return strconv.FormatBool(v)
	default:
		return fmt.Sprint(v)
	}
}

// number narrows a JSON scalar to a finite float. A non-finite value (the engine
// prints a bare `nan` for a channel it could not measure) is NOT a measurement,
// so it reads as absent.
func number(raw any) (float64, bool) {
	switch v := raw.(type) {
	case float64:
		return v, !math.IsNaN(v) && !math.IsInf(v, 0)
	case bool:
		if v {
			return 1, true
		}
		return 0, true
	case string:
		f, err := strconv.ParseFloat(v, 64)
		return f, err == nil && !math.IsNaN(f) && !math.IsInf(f, 0)
	}
	return 0, false
}

// truthy reads a boolean field. The CLI spells one as a JSON bool, or as the
// classifier that decided it (`"use": "def"` behind `reference.defines`).
func truthy(raw any) bool {
	switch v := raw.(type) {
	case bool:
		return v
	case string:
		switch v {
		case "def", "true", "yes", "verified":
			return true
		}
		return false
	case float64:
		return v != 0
	}
	return false
}

// ndjson decodes a verb's stdout — one JSON object per non-empty line — and
// withholds the summary record. Every face marks that record with a "verb" key
// and no row schema declares such a field, so the discriminator is the CLI's own
// convention rather than a position: pack prints its summary last, echoes first.
func ndjson(stdout string) ([]map[string]any, error) {
	var out []map[string]any
	for line := range strings.SplitSeq(stdout, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || line[0] != '{' {
			continue
		}
		var obj map[string]any
		if err := json.Unmarshal(finite([]byte(line)), &obj); err != nil {
			return nil, fmt.Errorf("irregex: undecodable row from the engine: %w", err)
		}
		if _, isSummary := obj["verb"]; !isSummary {
			out = append(out, obj)
		}
	}
	return out, nil
}

// summary is the verb's own account of the answer, read from stdout where the
// faces print it now and from stderr for a binary that still prints it there.
func summary(stdout, stderr string) map[string]any {
	for line := range strings.SplitSeq(stdout, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || line[0] != '{' {
			continue
		}
		var obj map[string]any
		if json.Unmarshal(finite([]byte(line)), &obj) == nil {
			if _, ok := obj["verb"]; ok {
				return obj
			}
		}
	}
	return diagnostic(stderr)
}

// finite rewrites the bare `nan` / `inf` tokens the engine prints for an
// unmeasurable channel into JSON `null`. They are not JSON, and a strict decoder
// would reject the whole row over a field the answer does not depend on; null
// carries the same meaning the presence mask does — not measured.
func finite(line []byte) []byte {
	var out []byte
	for i, inString := 0, false; i < len(line); {
		switch c := line[i]; {
		case c == '"':
			inString = !inString
			out, i = append(out, c), i+1
		case c == '\\' && inString && i+1 < len(line):
			out, i = append(out, c, line[i+1]), i+2
		case !inString && (c == 'n' || c == 'i' || c == '-'):
			if n := nonFinite(line[i:]); n > 0 {
				out, i = append(out, "null"...), i+n
				continue
			}
			out, i = append(out, c), i+1
		default:
			out, i = append(out, c), i+1
		}
	}
	return out
}

func nonFinite(s []byte) int {
	for _, token := range [...]string{"-nan", "-inf", "-infinity", "nan", "inf", "infinity"} {
		if len(s) >= len(token) && strings.EqualFold(string(s[:len(token)]), token) {
			return len(token)
		}
	}
	return 0
}

// diagnostic is the verb's stderr summary record under --json — how large a
// population the answer was drawn from, and whether it came warm from a persisted
// artifact or from a live build.
func diagnostic(stderr string) map[string]any {
	for _, raw := range slices.Backward(strings.Split(stderr, "\n")) {
		line := strings.TrimSpace(raw)
		if line == "" || line[0] != '{' {
			continue
		}
		var obj map[string]any
		if json.Unmarshal(finite([]byte(line)), &obj) == nil {
			return obj
		}
	}
	return nil
}
